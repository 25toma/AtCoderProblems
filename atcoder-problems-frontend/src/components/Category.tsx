import * as React from "react";
import { Table, Grid } from "react-bootstrap";
import {} from "react-bootstrap";
import { CategoryOneBlock } from "./CategoryOneBlock";
import { CategorySmallBlock } from "./CategorySmallBlock";
import { Problem } from "../model/Problem";
import { Contest } from "../model/Contest";

export interface CategoryProps {
  problems: Array<Problem>;
  contests: Array<Contest>;
  userId: string;
  rivals: Array<String>;
}

export class Category extends React.Component<CategoryProps, {}> {
  private filterProblems(
    problemMap: Map<string, Array<Problem>>,
    regexp: RegExp
  ): Array<[Contest, Array<Problem>]> {
    return this.props.contests
      .filter(contest => contest.id.match(regexp))
      .sort((a, b) => a.id.localeCompare(b.id))
      .reverse()
      .filter(contest => problemMap.has(contest.id))
      .map(contest => {
        let problems = problemMap
          .get(contest.id)
          .sort((a, b) => a.id.localeCompare(b.id));
        let r: [Contest, Array<Problem>] = [contest, problems];
        return r;
      });
  }

  render() {
    let problemMap = new Map<string, Array<Problem>>();
    this.props.contests.forEach(contest => problemMap.set(contest.id, []));
    this.props.problems
      .filter(problem => problemMap.has(problem.contestId))
      .forEach(problem => problemMap.get(problem.contestId).push(problem));

    // filter problems of AtCoder official contests
    let agc = this.filterProblems(problemMap, /^agc\d{3}$/);
    let abc = this.filterProblems(problemMap, /^abc\d{3}$/);
    let arc = this.filterProblems(problemMap, /^arc\d{3}$/);

    // problems of other contests
    let others = this.filterProblems(problemMap, /^(?!a[rgb]c\d{3}).*$/).sort(
      (a, b) => {
        return b[0].start_epoch_second - a[0].start_epoch_second;
      }
    );

    return (
      <Grid>
        <CategoryOneBlock
          categoryTitle="AtCoder Grand Contest"
          data={agc}
          header={"ABCDEF".split("")}
        />
        <CategoryOneBlock
          categoryTitle="AtCoder Beginner Contest"
          data={abc}
          header={"ABCD".split("")}
        />
        <CategoryOneBlock
          categoryTitle="AtCoder Regular Contest"
          data={arc}
          header={"ABCD".split("")}
        />
        <CategorySmallBlock data={others} />
      </Grid>
    );
  }
}
