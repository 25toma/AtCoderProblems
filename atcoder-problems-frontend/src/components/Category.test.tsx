import * as React from "react";
import { Category } from "./Category";
import { shallow } from "enzyme";
import { configure } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

configure({ adapter: new Adapter() });
test("split problems", () => {
  let problems = [
    { contestId: "arc999", id: "problem1", title: "" },
    { contestId: "agc999", id: "problem2", title: "" },
    { contestId: "abc999", id: "problem3", title: "" },
    { contestId: "other", id: "problem4", title: "" }
  ];
  let contests = [
    {
      id: "arc999",
      title: "AtCoder Regular Contest 999",
      start_epoch_second: 0
    },
    {
      id: "abc999",
      title: "AtCoder Beginner Contest 999",
      start_epoch_second: 0
    },
    {
      id: "agc999",
      title: "AtCoder Grand Contest 999",
      start_epoch_second: 0
    },
    {
      id: "other",
      title: "AtCoder Other Contest",
      start_epoch_second: 0
    }
  ];
  let wrapper = shallow(
    <Category problems={problems} contests={contests} userId={""} rivals={[]} />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
