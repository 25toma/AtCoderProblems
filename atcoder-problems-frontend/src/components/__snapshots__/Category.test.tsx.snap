// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`split problems 1`] = `
<Grid
  bsClass="container"
  componentClass="div"
  fluid={false}
>
  <CategoryOneBlock
    categoryTitle="AtCoder Grand Contest"
    data={
      Array [
        Array [
          Object {
            "id": "agc999",
            "start_epoch_second": 0,
            "title": "AtCoder Grand Contest 999",
          },
          Array [
            Object {
              "contestId": "agc999",
              "id": "problem2",
              "title": "",
            },
          ],
        ],
      ]
    }
    header={
      Array [
        "A",
        "B",
        "C",
        "D",
        "E",
        "F",
      ]
    }
  />
  <CategoryOneBlock
    categoryTitle="AtCoder Beginner Contest"
    data={
      Array [
        Array [
          Object {
            "id": "abc999",
            "start_epoch_second": 0,
            "title": "AtCoder Beginner Contest 999",
          },
          Array [
            Object {
              "contestId": "abc999",
              "id": "problem3",
              "title": "",
            },
          ],
        ],
      ]
    }
    header={
      Array [
        "A",
        "B",
        "C",
        "D",
      ]
    }
  />
  <CategoryOneBlock
    categoryTitle="AtCoder Regular Contest"
    data={
      Array [
        Array [
          Object {
            "id": "arc999",
            "start_epoch_second": 0,
            "title": "AtCoder Regular Contest 999",
          },
          Array [
            Object {
              "contestId": "arc999",
              "id": "problem1",
              "title": "",
            },
          ],
        ],
      ]
    }
    header={
      Array [
        "A",
        "B",
        "C",
        "D",
      ]
    }
  />
  <CategorySmallBlock
    data={
      Array [
        Array [
          Object {
            "id": "other",
            "start_epoch_second": 0,
            "title": "AtCoder Other Contest",
          },
          Array [
            Object {
              "contestId": "other",
              "id": "problem4",
              "title": "",
            },
          ],
        ],
      ]
    }
  />
</Grid>
`;
